File: news_server/modules/source/infrastructure/getters/steam.ts
/**
 * Steam 新闻源获取器
 * 从 newsnow/server/sources/steam.ts 迁移
 */
import * as cheerio from 'cheerio'
import type { SourceGetter } from '../../domain/getter/getter'
import { NewsItem } from '../../domain/newsitem/entity'

export const steamGetter: SourceGetter = async () => {
  const response = await fetch('https://store.steampowered.com/stats/stats/')
  
  if (!response.ok) {
    throw new Error(`Failed to fetch Steam: ${response.statusText}`)
  }

  const html = await response.text()
  const $ = cheerio.load(html)
  
  const news: NewsItem[] = []
  const $rows = $('#detailStats tr.player_count_row')

  $rows.each((_, el) => {
    const $el = $(el)
    const $a = $el.find('a.gameLink')
    const url = $a.attr('href')
    const gameName = $a.text().trim()
    const currentPlayers = $el.find('td:first-child .currentServers').text().trim()

    if (url && gameName && currentPlayers) {
      news.push(
        NewsItem.fromEntity({
          id: url,
          title: gameName,
          url,
          pubDate: Date.now(),
          extra: {
            info: currentPlayers,
          },
        })
      )
    }
  })

  return news
}

File: modules/source/infrastructure/getters/steam.go
package getters

import (
	"context"
	"strings"

	"github.com/PuerkitoBio/goquery"
)

func (r *Registry) steam(ctx context.Context) ([]Item, error) {
	body, err := r.get(ctx, "https://store.steampowered.com/stats/stats/", nil)
	if err != nil {
		return nil, err
	}
	doc, err := goquery.NewDocumentFromReader(strings.NewReader(string(body)))
	if err != nil {
		return nil, err
	}
	var out []Item
	doc.Find("#detailStats tr.player_count_row").Each(func(_ int, s *goquery.Selection) {
		a := s.Find("a.gameLink")
		href, _ := a.Attr("href")
		name := strings.TrimSpace(a.Text())
		players := strings.TrimSpace(s.Find("td:first-child .currentServers").Text())
		if href == "" || name == "" {
			return
		}
		out = append(out, Item{ID: href, Title: name, URL: href, Extra: map[string]any{"info": players}})
	})
	return out, nil
}
